acra-censor/handlers/handlers_util.go

```diff
@@ -548,51 +596,274 @@ func IsEqualComparisonNodes(patternNode, queryNode *sqlparser.ComparisonExpr) bo
+		return true
+	}
+
+	return false
+}
+
+// isEqualComparisonNodes try to match patternNode with queryNode with supported patterns for ComparisonExpr
+func IsEqualComparisonNodes(patternNode, queryNode *sqlparser.ComparisonExpr, whereType string) bool {
+	switch whereType {
+	case sqlparser.HavingStr:
+		return isEqualComparisonNodesInHaving(patternNode, queryNode)
+	case sqlparser.WhereStr:
+		return isEqualComparisonNodesInWhere(patternNode, queryNode)
+	default:
+		// unexpected type of parent node
+		return false
+	}
+}
+
+// matchValue return true if pattern node is %%VALUE%% pattern and value of query node has type that masked with this pattern
+func matchValue(patternNode, queryNode sqlparser.SQLNode) bool {
+	return isValuePattern(patternNode) && matchSQLValue(queryNode)
+}
+
+// matchSubquery return true if pattern are %%SUBQUERY%% and queryNode has correct type for this pattern otherwise false
+func matchSubquery(patternNode, queryNode sqlparser.SQLNode) bool {
+	if _, ok := patternNode.(*sqlparser.Subquery); !ok {
+		return false
+	}
+	if _, ok := queryNode.(*sqlparser.Subquery); !ok {
+		return false
+	}
+	// check that patterns query the same as our parsed placeholder
+	if reflect.DeepEqual(patternNode.(*sqlparser.Subquery).Select, SubqueryConfigPlaceholderReplacerParsed) {
+		return true
+	}
+	return false
+}
+
+// matchSQLValue return true if %%VALUE%% pattern should mask value of node
+// return true for any literal values, boolean and null
+// return false on other values like subqueries
+func matchSQLValue(node sqlparser.SQLNode) bool {
+	switch node.(type) {
+	case *sqlparser.SQLVal, sqlparser.BoolVal, *sqlparser.NullVal:
+		return true
+	}
+	return false
+}
+
+// matchRangeCondition handle range queries (age BETWEEN %%value%% and 5)
+// return true if match (with or without %%value%% patterns) otherwise false
+func matchRangeCondition(patternNode, queryNode *sqlparser.RangeCond) bool {
+	if queryNode.Operator != patternNode.Operator {
+		return false
+	}
+	if !reflect.DeepEqual(queryNode.Left, patternNode.Left) {
+		return false
+	}
+	if !(matchValue(patternNode.From, queryNode.From) || matchSubquery(patternNode.From, queryNode.From)) {
+		if !reflect.DeepEqual(patternNode.From, queryNode.From) {
+			return false
 		}
-		if queryWhereNode, ok := queryNodes[index+queryNodeOffset].(*sqlparser.Where); ok && queryWhereNode != nil {
-			queryWhereDetected = true
+	}
+	if !(matchValue(patternNode.To, queryNode.To) || matchSubquery(patternNode.To, queryNode.To)) {
+		if !reflect.DeepEqual(patternNode.To, queryNode.To) {
+			return false
 		}
-		if queryWhereDetected && patternWhereDetected {
-			return true
+	}
+	return true
+}
+
+// matchOrderBy handles order by construction
+// return true if match otherwise false
+func matchOrderBy(patternNode, queryNode sqlparser.OrderBy) bool {
+	if reflect.DeepEqual(patternNode, queryNode) {
+		return true
+	}
+
+	if len(patternNode) != len(queryNode) {
```

Comment:
we don't need the check of length because in a future here may be list of columns or values pattern and length of pattern nodes will not equal to length of query nodes

Reply:
Lets remove it when we infroduce such placeholder. Now it seems correct to check that amounts of column placeholders equals to actual amount of columns in query